integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/VisitorFactory.java

```diff
@@ -0,0 +1,23 @@
+package io.openlineage.spark.agent.lifecycle.plan;
+
+import io.openlineage.client.OpenLineage;
+import java.util.List;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import scala.PartialFunction;
+
+/**
+ * Provides Visitors for iterating on {@link LogicalPlan}.
+ *
+ * <p>All common Visitors should be grouped and passed to {@link
+ * VisitorFactory#getOutputVisitors(List)} to retrieve Visitors for {@link
+ * OpenLineage.OutputDataset}
+ */
+public interface VisitorFactory {
+
+  List<PartialFunction<LogicalPlan, List<OpenLineage.Dataset>>> getCommonVisitors();
```

Comment:
Can the interface just include `getInputVisitors` and `getOutputVisitors` and the common visitors can just be an internal implementation detail?

Reply:
not realy, only if we redesign Visitors approach, cause we should get all common visitors and pass them to output visitors, so in our case, version specific visitors (`DatasetSourceVisitor`) should be passed to common output visitors (like to `InsertIntoDataSourceVisitor`)

Reply:
Sorry, not sure which of your statements came first. 

The implementation of `getInputVisitors` could be 
```java
List<PartialFunction<LogicalPlan, List<OpenLineage.Dataset>>> visitors = new ArrayList<>(getCommonVisitors());
visitors.addAll(inputVisitors);
return visitors;
```
Then the `getCommonVisitors` is an internal implementation detail, but not exposed in the interface.